lib/includes.js

```diff
@@ -0,0 +1,189 @@
+var Promise = require('bluebird'),
+    _ = require('lodash'),
+    $http = require('http-as-promised');
+
+module.exports = function (adapter, schemas) {
+
+    function appendLinkedResources(body, primaryResourceName, inclusions) {
+
+        var primarySchema = schemas[primaryResourceName];
+
+        function isRemote(branch, key) {
+            return null != branch && null != branch.refs && null != branch.refs[key] && null != branch.refs[key].def && null != branch.refs[key].def.baseUri;
+        }
+
+        function toIncludes(map) {
+            var includes = [];
+
+            function buildRecursively(map, prefix) {
+                prefix = prefix || '';
+                return _.forEach(map, function (value, key) {
+                    if (_.isEmpty(value)) {
+                        includes.push(prefix + key);
+                    } else {
+                        var newPrefix = prefix + key + '.';
+                        buildRecursively(value, newPrefix)
+                    }
+                });
+            }
+
+            buildRecursively(map);
+            return includes.join(',');
+        }
+
+        function mapToInclusionTree(typeMap, schema, path) {
+            path = path || [];
+            var branch = _.map(typeMap, function (value, key) {
+                path.push(key);
+                var subBranch = buildInclusionBranch([key], schema);
+                if (!_.isEmpty(value)) {
+                    if (isRemote(subBranch, key)) {
+                        subBranch.refs[key].def.remoteIncludes = toIncludes(value);
+                    } else {
+                        var schemaName = schema.relationships[key].type;
+                        var subMap = mapToInclusionTree(value, schemas[schemaName], path);
+                        _.merge(subBranch.refs[key], subMap);
+                    }
+                }
+                path.pop();
+                return subBranch;
+            });
+            return _.reduce(branch, function (acc, item) {
+                return _.merge(acc, item);
+            }, {});
+        }
+
+        var inclusionMap = _.map(inclusions, function (inclusionString) {
```

Comment:
such utility functions can be moved to a ```util``` module maybe? It's be nice if we only had the core functions here. Things that describe themselves when you read the code. Art of storytelling with code :smile: